File: clients/python/devlog_client/entry.py
# Pure builders for the devicelog.v1 wire types. These construct real protobuf
# LogEntry messages from the generated stubs but perform no I/O, so they are
# trivially unit-testable and their output serializes byte-for-byte the same as
# every other language client (one schema, no drift).
from __future__ import annotations

import time
from typing import Dict, List, Mapping, Optional, Tuple

from devlog_client.gen.devicelog.v1 import log_pb2

# Mirror of the devicelog.v1 enums by their proto numeric values. The generated
# module exposes the SEVERITY_* names; these aliases match the Node client's
# short spellings for readability and cross-client parity.
Severity = log_pb2.Severity
SanitizationPhase = log_pb2.SanitizationPhase
SanitizationStandard = log_pb2.SanitizationStandard

# Short aliases (numeric values are identical to the proto enum entries).
SEVERITY_INFO = Severity.SEVERITY_INFO
SEVERITY_WARN = Severity.SEVERITY_WARN


def topic_for(device_id: str, subsystem: str) -> str:
    """The MQTT topic a device publishes to. The broker's ACL pins a session to
    devlog/v1/<its-own-id>/#, so device must equal the authenticated identity."""
    return f"devlog/v1/{device_id}/{subsystem}"


def to_timestamp(seconds_float: Optional[float] = None):
    """google.protobuf.Timestamp from epoch seconds (or now). Split into whole
    seconds + nanos exactly like the Node client's millis math."""
    if seconds_float is None:
        seconds_float = time.time()
    ms = int(round(seconds_float * 1000))
    from google.protobuf.timestamp_pb2 import Timestamp

    ts = Timestamp()
    ts.seconds = ms // 1000
    ts.nanos = (ms % 1000) * 1_000_000
    return ts


def build_log_entry(
    *,
    device_id: str,
    subsystem: str,
    seq: int = 0,
    severity: int = None,
    message: str = "",
    trace_id: str = "",
    device_time: Optional[float] = None,
    attributes: Optional[Mapping[str, str]] = None,
    payload: Optional[bytes] = None,
    sanitization: Optional[log_pb2.SanitizationEvent] = None,
) -> log_pb2.LogEntry:
    """Build a LogEntry protobuf message. entry_id, ingest_time and audit are
    deliberately never set — the server owns them and rejects producer-supplied
    values."""
    if not device_id:
        raise ValueError("deviceId is required")
    if not subsystem:
        raise ValueError("subsystem is required")
    if severity is None:
        severity = Severity.SEVERITY_INFO

    e = log_pb2.LogEntry(
        device_id=device_id,
        seq=seq,
        severity=severity,
        subsystem=subsystem,
        message=message,
        trace_id=trace_id,
    )
    e.device_time.CopyFrom(to_timestamp(device_time))
    if attributes:
        for k, v in attributes.items():
            e.attributes[k] = v
    if payload:
        e.payload = payload
    if sanitization is not None:
        e.sanitization.CopyFrom(sanitization)
    return e


def build_sanitization_job(device_id: str, trace_id: str) -> Tuple[str, List[log_pb2.LogEntry]]:
    """A complete, ordered sanitization job — the same sequence the C++ and Node
    reference producers emit: STARTED -> 3x PROGRESS -> VERIFYING -> COMPLETED.
    Returns (trace_id, entries) so callers can publish then query it back."""
    media = log_pb2.TargetMedia(
        serial="WD-9F2K3L0042",
        model="WDC WD40EFRX",
        capacity_bytes=4 * 2 ** 40,  # 4 TiB
        media_type="HDD",
    )

    seq = 0

    def step(phase: int, message: str, *, progress_pct: float = None, verification=None) -> log_pb2.LogEntry:
        nonlocal seq
        seq += 1
        san = log_pb2.SanitizationEvent(
            media=media,
            standard=SanitizationStandard.SANITIZATION_STANDARD_NIST_800_88_PURGE,
            technique="overwrite-1pass",
            operator_id="op-python",
            phase=phase,
        )
        if progress_pct is not None:
            san.progress_pct = progress_pct
        if verification is not None:
            san.verification.CopyFrom(verification)
        return build_log_entry(
            device_id=device_id,
            seq=seq,
            severity=Severity.SEVERITY_INFO,
            subsystem="sanitizer",
            message=message,
            trace_id=trace_id,
            sanitization=san,
        )

    P = SanitizationPhase
    entries = [
        step(P.SANITIZATION_PHASE_STARTED, "sanitization started"),
        step(P.SANITIZATION_PHASE_PROGRESS, "overwrite pass in progress", progress_pct=25),
        step(P.SANITIZATION_PHASE_PROGRESS, "overwrite pass in progress", progress_pct=50),
        step(P.SANITIZATION_PHASE_PROGRESS, "overwrite pass in progress", progress_pct=75),
        step(P.SANITIZATION_PHASE_VERIFYING, "verifying sampled sectors", progress_pct=100),
        step(
            P.SANITIZATION_PHASE_COMPLETED,
            "sanitization completed, verification passed",
            progress_pct=100,
            verification=log_pb2.Verification(method="sampled-read", sample_pct=10, passed=True),
        ),
    ]
    return trace_id, entries

File: clients/nest/src/devlog/entry.ts
// Pure, dependency-free builders for the devicelog.v1 wire types. Kept free of
// I/O and protobuf runtime so they are trivially unit-testable; proto.ts turns
// the plain objects these produce into encoded bytes.

// Mirror of the devicelog.v1 enums (proto numeric values).
export const Severity = {
  UNSPECIFIED: 0, TRACE: 1, DEBUG: 2, INFO: 3, WARN: 4, ERROR: 5, FATAL: 6,
} as const;

export const SanitizationPhase = {
  UNSPECIFIED: 0, STARTED: 1, PROGRESS: 2, VERIFYING: 3,
  COMPLETED: 4, FAILED: 5, ABORTED: 6,
} as const;

export const SanitizationStandard = {
  UNSPECIFIED: 0,
  NIST_800_88_CLEAR: 1, NIST_800_88_PURGE: 2, NIST_800_88_DESTROY: 3,
  IEEE_2883_CLEAR: 4, IEEE_2883_PURGE: 5, IEEE_2883_DESTRUCT: 6,
} as const;

export interface Timestamp {
  seconds: number;
  nanos: number;
}

// google.protobuf.Timestamp from a JS Date (or now).
export function toTimestamp(date: Date = new Date()): Timestamp {
  const ms = date.getTime();
  return { seconds: Math.floor(ms / 1000), nanos: (ms % 1000) * 1e6 };
}

// The MQTT topic a device publishes to. The broker's ACL pins a session to
// devlog/v1/<its-own-id>/#, so device must equal the authenticated identity.
export function topicFor(deviceId: string, subsystem: string): string {
  return `devlog/v1/${deviceId}/${subsystem}`;
}

export interface LogEntry {
  deviceId: string;
  seq?: number;
  severity: number;
  subsystem: string;
  message?: string;
  traceId?: string;
  deviceTime: Timestamp;
  attributes?: Record<string, string>;
  payload?: Uint8Array;
  sanitization?: Record<string, unknown>;
}

export interface BuildLogEntryArgs {
  deviceId: string;
  seq?: number;
  severity?: number;
  subsystem: string;
  message?: string;
  traceId?: string;
  deviceTime?: Date;
  attributes?: Record<string, string>;
  payload?: Uint8Array;
  sanitization?: Record<string, unknown>;
}

// Build a LogEntry as a plain object using protobufjs camelCase field names.
// entry_id, ingest_time and audit are deliberately omitted — the server owns
// them and rejects producer-supplied values.
export function buildLogEntry({
  deviceId,
  seq,
  severity = Severity.INFO,
  subsystem,
  message,
  traceId,
  deviceTime = new Date(),
  attributes,
  payload,
  sanitization,
}: BuildLogEntryArgs): LogEntry {
  if (!deviceId) throw new Error('deviceId is required');
  if (!subsystem) throw new Error('subsystem is required');
  const e: LogEntry = {
    deviceId,
    seq,
    severity,
    subsystem,
    message,
    traceId,
    deviceTime: toTimestamp(deviceTime),
  };
  if (attributes) e.attributes = attributes;
  if (payload) e.payload = payload;
  if (sanitization) e.sanitization = sanitization;
  return e;
}

export interface SanitizationJob {
  traceId: string;
  entries: LogEntry[];
}

// A complete, ordered sanitization job — the same sequence the C++ reference
// producer and `logctl sim` emit: STARTED -> 3x PROGRESS -> VERIFYING ->
// COMPLETED. Returns { traceId, entries } so callers can publish then query it.
export function buildSanitizationJob(deviceId: string, traceId: string): SanitizationJob {
  const media = {
    serial: 'WD-9F2K3L0042',
    model: 'WDC WD40EFRX',
    capacityBytes: 4 * 2 ** 40,
    mediaType: 'HDD',
  };
  const base = {
    media,
    standard: SanitizationStandard.NIST_800_88_PURGE,
    technique: 'overwrite-1pass',
    operatorId: 'op-nest',
  };
  let seq = 0;
  const step = (phase: number, message: string, extra: Record<string, unknown> = {}) =>
    buildLogEntry({
      deviceId,
      seq: ++seq,
      severity: Severity.INFO,
      subsystem: 'sanitizer',
      message,
      traceId,
      sanitization: { ...base, phase, ...extra },
    });

  const entries = [
    step(SanitizationPhase.STARTED, 'sanitization started'),
    step(SanitizationPhase.PROGRESS, 'overwrite pass in progress', { progressPct: 25 }),
    step(SanitizationPhase.PROGRESS, 'overwrite pass in progress', { progressPct: 50 }),
    step(SanitizationPhase.PROGRESS, 'overwrite pass in progress', { progressPct: 75 }),
    step(SanitizationPhase.VERIFYING, 'verifying sampled sectors', { progressPct: 100 }),
    step(SanitizationPhase.COMPLETED, 'sanitization completed, verification passed', {
      progressPct: 100,
      verification: { method: 'sampled-read', samplePct: 10, passed: true },
    }),
  ];
  return { traceId, entries };
}
